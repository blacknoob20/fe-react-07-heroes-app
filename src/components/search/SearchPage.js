import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchPage = ({ history }) => {
    const location = useLocation();
    // const parsed = queryString.parse(location.search);
    const { q = '' } = queryString.parse(location.search);
    const [{ search }, handleInputChange] = useForm({ search: q });
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`);
    }

    return (
        <div>
            <h1>Search Page</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input type="text" name="search" onChange={handleInputChange} value={search} placeholder="Find your hero" className="form-control" autoComplete="off" />
                        <button type="submit" className="btn btn-outline-primary m-1">Search...</button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        heroesFiltered.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
