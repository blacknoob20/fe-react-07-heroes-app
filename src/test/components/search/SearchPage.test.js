import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchPage } from '../../../components/search/SearchPage';

describe('Pruebas en <SearchPage/>', () => {
    test('Debe mostrarse correctamente, con valores por defecto.', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/']}>
                <Route path="/search" component={SearchPage} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
    });
    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchPage} />
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });
    test('Debe de mostrar un error si no se encuentra el Hero.', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=1321313456465']}>
                <Route path="/search" component={SearchPage} />
            </MemoryRouter>
        );
        expect(wrapper.find('p').text()).toBe('No hay resultados');
    });
    test('Debe de llamar el push del history.', () => {
        const history = {
            push: jest.fn()
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route
                    path="/search"
                    component={() => <SearchPage history={history} />}
                />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change', {
            target: {
                nanme: 'searchtext',
                value: 'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        expect(history.push).toHaveBeenCalledWith(`?q=batman`);
    });

});
