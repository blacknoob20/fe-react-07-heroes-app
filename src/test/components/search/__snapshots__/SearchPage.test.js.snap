// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> Debe mostrarse correctamente, con valores por defecto. 1`] = `""`;
